### PR TITLE
std: Stabilize APIs for the 1.12 release

### LINKED ISSUE #27708: Tracking issue for `as_unsafe_cell` stabilization
The `Cell::as_unsafe_cell` and `RefCell::as_unsafe_cell` methods return a reference to the underlying `UnsafeCell` in both cases.

It's not clear what the use cases are for these methods and whether they can be achieved in some alternative way. If you are using these methods, please leave a comment with your use case.

### LINKED ISSUE #32313: Tracking issue for Ipv6Addr octet methods
Tracking issue for https://github.com/rust-lang/rfcs/pull/1498

First needs to be implemented, then go through an FCP!

Looks like the `From<[u8; 4]> for Ipv4Addr` is being implemented over at https://github.com/rust-lang/rust/pull/32050.

### LINKED ISSUE #32630: Tracking issue for {VecDeque,LinkedList}::contains
Tracking issue for https://github.com/rust-lang/rfcs/pull/1552

### LINKED ISSUE #32713: process::ExitStatus should have some kind of constructor
Because `ExitStatus` has a private field it is impossible to create one in user code, one must actually execute a process. This prevents testing/mocking. My use case is that I have some function which runs a process and generates data, however, for testing I want to mock the test run and just return a success exit status and some data. I can't do that because I can't create my own `ExitStatus` objects.

At the least there should be a way to create a success `ExitStatus`. Preferably, there would be a way to create various fail statuses too.

### LINKED ISSUE #34029: Tracking issue for Receiver::recv_timeout
Added in https://github.com/rust-lang/rust/pull/33748, this is a tracking issue for having a timeout on receiving on a channel.

### LINKED ISSUE #34285: Tracking for map_entry_recover_keys

### LINKED ISSUE #34392: Tracking issue for BinaryHeap::peek_mut
cc rust-lang/rfcs#1626 @alexcrichton

### LINKED ISSUE #34529: Tracking issue for iter::{Product, Sum} traits
This is a tracking issue for the recently-added `iter::Product` and `iter::Sum` traits which are used as the implementation of the `Iterator::product` and `Iterator::sum` methods.

### PR DESCRIPTION
**Pull request by alexcrichton**
Stabilized
- `Cell::as_ptr`
- `RefCell::as_ptr`
- `IpAddr::is_{unspecified,loopback,multicast}`
- `Ipv6Addr::octets`
- `LinkedList::contains`
- `VecDeque::contains`
- `ExitStatusExt::from_raw` - both on Unix and Windows
- `Receiver::recv_timeout`
- `RecvTimeoutError`
- `BinaryHeap::peek_mut`
- `PeekMut`
- `iter::Product`
- `iter::Sum`
- `OccupiedEntry::remove_entry`
- `VacantEntry::into_key`

Deprecated
- `Cell::as_unsafe_cell`
- `RefCell::as_unsafe_cell`
- `OccupiedEntry::remove_pair`

Closes #27708
cc #27709
Closes #32313
Closes #32630
Closes #32713
Closes #34029
Closes #34392
Closes #34285
Closes #34529

### CODE REVIEW

**Comment by alexcrichton** [comment]
r? @aturon 

cc @rust-lang/libs

File: src/libstd/net/ip.rs [hunk 1]
```diff
@@ -86,8 +85,6 @@ impl IpAddr {
     /// Returns true if the address appears to be globally routable ([IPv4], [IPv6]).
     /// [IPv4]: ../../std/net/struct.Ipv4Addr.html#method.is_global
     /// [IPv6]: ../../std/net/struct.Ipv6Addr.html#method.is_global
-    #[unstable(feature="ip", issue="27709",
-               reason="recently added and depends on unstable Ip{v4,v6}Addr.is_global()")]
     pub fn is_global(&self) -> bool {
```
Comment (brson) [inline src/libstd/net/ip.rs]: Why do this and `is_documentation` have no tag at all now?

**Comment by brson** [comment]
r=me assuming the question of the disappeared tags is resolved.

File: src/libstd/net/ip.rs (hunk 1 above)
Comment (alexcrichton) [inline src/libstd/net/ip.rs]: Ah yeah this file is tagged with `#![unstable]`, so it inherits from that.

**Pull request merged**